src/__tests__/parseWithPointers.spec.ts

```diff
@@ -514,4 +514,167 @@ european-cities: &cities
       expect(result.diagnostics).toEqual([]);
     });
   });
+
+  describe('invalid (not JSON-ish) mapping keys', () => {
+    const complex = `[2]: test
+{2:null}: false
+2: test`;
+
+    const responses = `responses:
+  "200": {}
+  400: {}
+  true: false
+  null: 2`;
+
+    it('always excludes any complex types', () => {
+      const { data: yamlData } = parseWithPointers(complex, { json: false });
+      const { data: jsonData } = parseWithPointers(complex);
+
+      expect(yamlData).toStrictEqual({
+        '2': 'test',
+      });
+
+      expect(yamlData).toStrictEqual(jsonData);
+    });
+
+    it('always includes all scalar mapping keys', () => {
+      const { data: yamlData } = parseWithPointers(responses, { json: false });
+      const { data: jsonData } = parseWithPointers(responses);
+
+      expect(yamlData).toStrictEqual({
+        responses: {
+          '200': {},
+          '400': {},
+          null: 2,
+          true: false,
+        },
+      });
+
+      expect(yamlData).toStrictEqual(jsonData);
+    });
+
+    it('warns about non-string scalar mapping keys', () => {
+      const { diagnostics } = parseWithPointers(responses);
+
+      expect(diagnostics).toEqual([
+        {
+          code: 'YAMLException',
+          message: 'mapping key must be a string scalar rather than number',
+          path: ['responses', '400'],
+          range: {
+            end: {
+              character: 5,
+              line: 2,
+            },
+            start: {
+              character: 2,
+              line: 2,
+            },
+          },
+          severity: DiagnosticSeverity.Error,
```

Comment:
This is technically valid YAML right? If so, should this be a warning? And maybe the `code` should not be `YAMLException` but rather some other code that indicates its more about being compliant with JSON than a YAML thing?

Same for rest of items in this PR that are more "warnings" about JSON compliance, than actual invalid yaml exceptions.

Reply:
> This is technically valid YAML right? If so, should this be a warning?

Yeah, true. That's perfectly valid YAML, so we can in fact change the error code.
Re severity, I agree with you. I can make it a warning in case of JSON parse option is set to warn / hint, and error otherwise.
Thoughts?

Reply:
Sounds good to me!